docs/streams/developer-guide/testing.html

```diff
@@ -0,0 +1,294 @@
+TopologyTestDriver testDriver = new TopologyTestDriver(topology, config);
+    </pre>
+    <p>
+    The test driver accepts <code>ConsumerRecord</code>s with key and value type <code>byte[]</code>.
+    Because <code>byte[]</code> types are hard to handle, the <code>ConsumerRecordFactory</code> can be used to generate those records providing regular Java types
+    for key and values as well as the corresponding serializers.
+    </p>
+    <pre>
+ConsumerRecordFactory&lt;String, Integer&gt; factory = new ConsumerRecordFactory&lt;&gt;("input-topic", new StringSerializer(), new IntegerSerializer());
+testDriver.pipe(factory.create("key", 42L));
+    </pre>
+    <p>
+    To verify the output, the test driver produces <code>ProducerRecord</code>s with key and value type <code>byte[]</code>.
+    For result verification, you can specify corresponding deserializers when reading the output record from the driver.
+    <pre>
+ProducerRecord&lt;String, Integer&gt; outputRecord = testDriver.readOutput("output-topic", new StringDeserializer(), new LongDeserializer());
+    </pre>
+    <p>
+    For result verification, you can use <code>OutputVerifier</code>.
+    It offers helper methods to compare only certain parts of the result record:
+    for example, you might only care about the key and value, but not the timestamp of the result record.
+    </p>
+    <pre>
+OutputVerifier.compareKeyValue(outputRecord, "key", 42L); // throws AssertionError if key or value does not match
+    </pre>
+    <p>
+    <code>TopologyTestDriver</code> supports punctuations, too.
+    Event-time punctuations are triggered automatically based on the processed records' timestamps.
+    Wall-clock-time punctuations can also be triggered by advancing the test driver's wall-clock-time (the driver mocks wall-clock-time internally to give users control over it).
+    </p>
+    <pre>
+testDriver.advanceWallClockTime(20L);
+    </pre>
+    </div>
+    <p>
+    Additionally, you can access state stores via the test driver before or after a test.
+    Accessing stores before a test is useful to pre-populate a store with some initial values.
+    After data was processed, expected updates to the store can be verified.
+    </p>
+    <pre>
+KeyValueStore store = testDriver.getKeyValueStore("store-name");
+    </pre>
+    <p>
+    Note, that you should always close the test driver at the end to make sure all resources are release properly.
+    </p>
+    <pre>
+testDriver.close();
+    </pre>
+
+    <h2>Example</h2>
+    <p>
+    Below, we show a complex example how to use the test driver and helper classes.
+    We create a topology that computed the maximum per key using a key-value-store.
+    While processing, no output is generated, but only the store is updated.
+    Output is only sent downstream based on event-time and wall-clock punctuations.
+    </p>
+    <pre>
+    @Test
+    public void processorShouldAggregateAndEmitOnPunctuation() {
```

Comment:
Wouldn't block for this, but it would be nice to have the test written in as we'd expect them to be written in streams, i.e., one test per thing we are testing. This could easily be broken into multiple tests, which i think would enhance readability.

Reply:
+1